### DISCUSSION TITLE
Clicking on either email_change_current or email_change_new link will remove both one time tokens, clicking on second links leads to an error

### DISCUSSION THREAD

**Discussion by maximilian-hammerl** [supabase/supabase#30017] [Questions]
Hi, I am trying to understand the flow how to change the e-mail address as a user.

If the user requests an e-mail address change, we generate the email_change_current and email_change_new links (see [Generate links to change current e-mail address](https://supabase.com/docs/reference/javascript/auth-admin-generatelink)) and send them to their current and new e-mail address respectively. After generating both links, the corresponding one time tokens are in the database, with correct user id, token types email_change_token_current and email_change_token_new and the token hashes match when comparing them to the links in the e-mails.

When the user clicks on either one of those links, the user will be logged in automatically with the expected access and refresh tokens in the URL fragment:

> `#access_token=e...&expires_at=1729546688&expires_in=3600&refresh_token=o...&token_type=bearer&type=email_change`

But both one time tokens are deleted from the database. When clicking on the other link, the user will not be logged in again. Instead we see this error message in the URL fragment:

> `#error=access_denied&error_code=403&error_description=Email+link+is+invalid+or+has+expired`

The e-mail address change is successful, even with this error message, but my first reaction was that something went wrong. My question is whether this is the intended behavior or whether we misunderstood something/implemented something incorrectly.

I would understand that the user has to login again after the successful e-mail change, but the error message is throwing us off.

**Comment by maximilian-hammerl** [comment]
We investigated this issue further and are even more confused. My earlier tests were all with my local setup, which includes the following in the `config.toml`:
```
# If enabled, a user will be required to confirm any email change on both the old, and new email
# addresses. If disabled, only the new email is required to confirm.
double_confirm_changes = true
```
After looking at this more closely, we noticed that clicking on either of the links will immediately confirm the e-mail address change and the new e-mail address will be active. Both one time tokens are removed and then clicking on the second link leads to the error message.

**The `double_confirm_changes` settings seems to be ignored, or there is another setting which we are not aware off.**

![image](https://github.com/user-attachments/assets/a6ca05be-b10b-4740-b8d3-9a3c1fa8fbe2)
But when we test this in our production environment with these settings, the following behavior is even more unexpected:

When we click on the first link, it does not log us in, but only the corresponding one time token is removed. When we click on the second link, it does log us in, the corresponding one time token is removed as well, and the e-mail address change is successful.
Important: It does not matter on which link we click first.

**Firstly, we would accept the behavior of our local setup be at least similar to the behavior of our production setup.
Secondly, the error message are less than clear and we are not able to find any detailed information how this flow should work correctly.**

**Comment by GaryAustin1** [comment]
Maybe this change:  https://github.com/supabase/auth/pull/1241 in Sept?

Are you on latest CLI version?

**Comment by maximilian-hammerl** [comment]
Yes, we are using the latest CLI version v1.207.9

**Comment by maximilian-hammerl** [comment]
If anyone stumbles across this discussion, I also created a bug report with more details and continued discussion: https://github.com/supabase/auth/issues/1820
